### PR TITLE
Feature/coal oven

### PR DESCRIPTION
**Pull request by lucasvdiepen**

### CODE REVIEW

File: Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+{
+    /// <summary>
+    /// A class responsible for controlling the coal oven.
+    /// </summary>
+    public class CoalOven : NetworkAbstractSingleton<CoalOven>
+    {
+        /// <summary>
+        /// Invoked when the coal amount changes.
+        /// </summary>
+        public Action<float> OnCoalAmountChanged;
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs]:
```suggestion
        public Action<float> onCoalAmountChanged;
```

Also move down to the other public declared vars

File: Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+            TrainEngine.Instance.SetEngineState(TrainEngineState.Active);
+        }
+
+        /// <summary>
+        /// Adds coal to the oven.
+        /// </summary>
+        /// <param name="amount">The amount of coal to add.</param>
+        [Server]
+        public void AddCoal(float amount)
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs]:
```suggestion
        public void AddCoal(float amount)
                => SetCoalAmount(Mathf.Min(CoalAmount + amount, _maxCoalAmount));
```

File: Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        {
+            SetCoalAmount(Mathf.Min(CoalAmount + amount, _maxCoalAmount));
+        }
+
+        /// <summary>
+        /// Removes coal from the oven.
+        /// </summary>
+        /// <param name="amount">The amount of coal to remove.</param>
+        [Server]
+        public void RemoveCoal(float amount)
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs]:
```suggestion
        public void RemoveCoal(float amount)
                => SetCoalAmount(Mathf.Max(CoalAmount - amount, 0));
```

File: Assets/Scripts/Framework/Gameplay/Interactions/Interactables/CoalOvenInteractable.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+using UnityEngine;
+
+using DerailedDeliveries.Framework.CoalOvenSystem;
+using DerailedDeliveries.Framework.Gameplay.Interactions.Grabbables;
+using DerailedDeliveries.Framework.Gameplay.Player;
+
+namespace DerailedDeliveries.Framework.Gameplay.Interactions.Interactables
+{
+    /// <summary>
+    /// A <see cref="Interactable"/> responsible for handling the coal oven.
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/Gameplay/Interactions/Interactables/CoalOvenInteractable.cs]:
```suggestion
    /// An <see cref="Interactable"/> responsible for handling the coal oven.
```

File: Assets/Scripts/Framework/Gameplay/Interactions/Grabbables/CoalGrabbable.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+using DerailedDeliveries.Framework.Gameplay.Interactions.Interactables;
+
+namespace DerailedDeliveries.Framework.Gameplay.Interactions.Grabbables
+{
+    /// <summary>
+    /// A <see cref="UseableGrabbable"/> responsible for handling the coal.
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/Gameplay/Interactions/Grabbables/CoalGrabbable.cs]:
```suggestion
    /// An <see cref="UseableGrabbable"/> responsible for handling the coal.
```

File: Assets/Scripts/Framework/Gameplay/Interactions/Interactables/CoalPileInteractable.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
+using FishNet.Object;
+using UnityEngine;
+
+using DerailedDeliveries.Framework.Gameplay.Interactions.Grabbables;
+using DerailedDeliveries.Framework.Gameplay.Player;
+
+namespace DerailedDeliveries.Framework.Gameplay.Interactions.Interactables
+{
+    /// <summary>
+    /// A <see cref="Interactable"/> responsible for handling the coal pile.
```
Comment (PiterGroot) [inline Assets/Scripts/Framework/Gameplay/Interactions/Interactables/CoalPileInteractable.cs]:
```suggestion
    /// An <see cref="Interactable"/> responsible for handling the coal pile.
```

**Review by PiterGroot — CHANGES_REQUESTED** [review]
Looking good.

File: Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs (hunk 1 above)
Comment (lucasvdiepen) [reply Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs]: The naming is correct according to the [convention](https://github.com/lucasvdiepen/DerailedDeliveries/wiki/Code-Conventions#events).

**Review by PiterGroot — APPROVED** [review]

File: Assets/Scripts/Framework/CoalOvenSystem/CoalBarPopup.cs [hunk 7]
```diff
@@ -0,0 +1,30 @@
+using UnityEngine;
+using UnityEngine.UI;
+
+using DerailedDeliveries.Framework.PopupManagement;
+
+namespace DerailedDeliveries.Framework.CoalOvenSystem
+{
+    /// <summary>
+    /// A <see cref="Popup"/> that handles the coal bar popup.
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/CoalOvenSystem/CoalBarPopup.cs]:
```suggestion
    /// A <see cref="Popup"/> class that handles the coal bar popup logic.
```

File: Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        {
+            CoalAmount = coalAmount;
+
+            if(IsServer)
+            {
+                if(CoalAmount <= 0.0001f)
+                {
+                    TrainEngine.Instance.SetEngineState(TrainEngineState.Inactive);
+                }
+            }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/CoalOvenSystem/CoalOven.cs]:
```suggestion
            if(IsServer && CoalAmount <= 0.0001f)
                TrainEngine.Instance.SetEngineState(TrainEngineState.Inactive);
```

File: Assets/Scripts/Framework/CoalOvenSystem/CoalPopup.cs [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace DerailedDeliveries.Framework.CoalOvenSystem
+{
+    /// <summary>
+    /// A <see cref="Popup"/> that handles the coal popup.
+    /// </summary>
+    public class CoalPopup : Popup
+    {
+        [SerializeField]
+        private float _lowCoalThreshold = 40f;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/CoalOvenSystem/CoalPopup.cs:14]:
```suggestion
        private float _coalThreshold = 40f;
```

File: Assets/Scripts/Framework/DamageRepairManagement/RepairPopup.cs [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+            if(health <= _lowHealthThreshold)
+            {
+                _repairPopupImage.sprite = _lowHealthSprite;
+                Show();
+                return;
+            }
+
+            Close();
+        }
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/DamageRepairManagement/RepairPopup.cs]:
```suggestion
        private void OnHealthChanged(int health)
        {
            if(health <= _criticalLowHealthThreshold)
            {
                UpdateAndShowPopup(_criticalLowHealthSprite);
                return;
            }
           
            if(health <= _lowHealthThreshold)
            {
                UpdateAndShowPopup(_lowHealthSprite);
                return;
            }
           
            Close();
        }
       
        private void UpdateAndShowPopup(Sprite newPopup)
        {
            _repairPopupImage.sprite = newPopup;
            Show();
        }
```

Comment (Joost-Robat) [reply Assets/Scripts/Framework/DamageRepairManagement/RepairPopup.cs]: Extra DRY.

File: Assets/Scripts/Framework/Gameplay/Interactions/Grabbables/CoalGrabbable.cs [hunk 11]
```diff
@@ -0,0 +1,13 @@
+using DerailedDeliveries.Framework.Gameplay.Interactions.Interactables;
+
+namespace DerailedDeliveries.Framework.Gameplay.Interactions.Grabbables
+{
+    /// <summary>
+    /// An <see cref="UseableGrabbable"/> responsible for handling the coal.
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Gameplay/Interactions/Grabbables/CoalGrabbable.cs]:
```suggestion
    /// A <see cref="UseableGrabbable"/> class responsible for handling the coal.
```

File: Assets/Scripts/Framework/Gameplay/Interactions/Interactables/CoalOvenInteractable.cs [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+using DerailedDeliveries.Framework.CoalOvenSystem;
+using DerailedDeliveries.Framework.Gameplay.Interactions.Grabbables;
+using DerailedDeliveries.Framework.Gameplay.Player;
+using DerailedDeliveries.Framework.DamageRepairManagement;
+
+namespace DerailedDeliveries.Framework.Gameplay.Interactions.Interactables
+{
+    /// <summary>
+    /// An <see cref="Interactable"/> responsible for handling the coal oven.
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Gameplay/Interactions/Interactables/CoalOvenInteractable.cs]:
```suggestion
    /// A <see cref="Interactable"/> class responsible for handling the coal oven.
```

File: Assets/Scripts/Framework/Gameplay/Interactions/Interactables/CoalPileInteractable.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
+using FishNet.Object;
+using UnityEngine;
+
+using DerailedDeliveries.Framework.Gameplay.Interactions.Grabbables;
+using DerailedDeliveries.Framework.Gameplay.Player;
+
+namespace DerailedDeliveries.Framework.Gameplay.Interactions.Interactables
+{
+    /// <summary>
+    /// An <see cref="Interactable"/> responsible for handling the coal pile.
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Gameplay/Interactions/Interactables/CoalPileInteractable.cs]:
```suggestion
    /// A <see cref="Interactable"/> class responsible for handling the coal pile.
```

File: Assets/Scripts/Framework/Train/TrainEngine.cs [hunk 14]
```diff
@@ -169,20 +168,9 @@ private void OnTrainDirectionChanged(bool newDirection)
...
+            OnEngineStateChanged?.Invoke(EngineState);
 
-                CurrentSpeedIndex = 0;
-                CurrentEngineAcceleration = _speedValues[CurrentSpeedIndex];
-            }
-            else
-            {
-                _friction = _startFriction;
-            }
+            _friction = newState == TrainEngineState.Inactive ? _standbyFriction : _startFriction;
```
Comment (Joost-Robat) [inline Assets/Scripts/Framework/Train/TrainEngine.cs:173]:
```suggestion
            _friction = newState == TrainEngineState.Inactive 
                ? _standbyFriction 
                : _startFriction;
```

**Review by Joost-Robat — CHANGES_REQUESTED** [review]
Some changes are required but overall very clean code and minimal comments.

File: Assets/Scripts/Framework/CoalOvenSystem/CoalPopup.cs (hunk 9 above)
Comment (lucasvdiepen) [reply Assets/Scripts/Framework/CoalOvenSystem/CoalPopup.cs:14]: Resolved.

**Review by Joost-Robat — APPROVED** [review]
Looks good! Ready for merge.

**Pull request merged**